### PR TITLE
[docs] Add more guideline on using ray in slurm cluster

### PR DESCRIPTION
**Pull request by pengzhenghao**
## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [x] This PR is not tested :(

### CODE REVIEW

**Comment by pengzhenghao** [comment]
By the way, I saw two push destinations `ray` and `bazel-ray`, I only push to `ray`. Is that correct?

**Review by sumanthratna — COMMENTED** [review]
thanks for the PR! 

more substantial comments coming later; busy right now

**Comment by pengzhenghao** [comment]
Is there anything I need to do?

**Comment by sumanthratna** [comment]
> Is there anything I need to do?

not right now, no! I'll take a better look at this PR on Saturday or Sunday and leave another review. Thanks again! @richardliaw and @amogkam might also want to take a quick look

**Comment by richardliaw** [comment]
Yes! We will take a look and merge by Christmas!

File: doc/source/cluster/slurm.rst [hunk 1]
```diff
@@ -92,3 +92,225 @@ Starter SLURM script
...
+
+
+There are other options you can use when calling `python launch.py`:
+
+* `--exp-name`: The experiment name. Will generate `{exp-name}_{date}-{time}.sh` and  `{exp-name}_{date}-{time}.log`.
+* `--command`: The command you wish to run. For example: `rllib train XXX` or `python XXX.py`.
+* `--num-gpus`: The number of GPUs you wish to use in each computing node. Default: 0.
+* `--node` (`-w`): The specify nodes you wish to use, in the same form of the return of `sinfo`. Automatically assign if not specify.
+* `--num-nodes` (`-n`): The number of nodes you wish to use. Default: 1.
+* `--partition` (`-p`): The partition you wish to use. Default: "chpc" (CUHK cluster partition name, change to yours!)
```
Comment (sumanthratna) [inline doc/source/cluster/slurm.rst:139]: Is chpc a common convention for naming slurm partitions? If not, maybe we should remove the default and make this a required flag

File: doc/source/cluster/slurm.rst [hunk 2]
```diff
@@ -92,3 +92,225 @@ Starter SLURM script
...
+        print(
+            "Job submitted! Script file is at: <{}>. Log file is at: <{}>".format(
+                script_file, "{}.log".format(job_name))
+        )
+        sys.exit(0)
+
+
+
+The advanced SBATCH template `sbatch_template.sh`
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (sumanthratna) [inline doc/source/cluster/slurm.rst:247]:
```suggestion
The advanced SBATCH template `sbatch_template.sh`
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by richardliaw** [comment]
Thanks! Btw, happy to take over this PR :)

File: doc/source/cluster/slurm.rst (hunk 1 above)
Comment (pengzhenghao) [reply doc/source/cluster/slurm.rst]: Now I have change the default partition to "", in which case the slurm will look for the default partition for the user.

File: doc/source/cluster/slurm.rst [hunk 3]
```diff
@@ -92,3 +92,225 @@ Starter SLURM script
...
+        with open(template_file, "r") as f:
+            text = f.read()
+        text = text.replace(JOB_NAME, job_name)
+        text = text.replace(NUM_NODES, str(args.num_nodes))
+        text = text.replace(NUM_GPUS_PER_NODE, str(args.num_gpus))
+        text = text.replace(PARTITION_NAME, str(args.partition))
+        text = text.replace(COMMAND_PLACEHOLDER, str(args.command))
+        text = text.replace(LOAD_ENV, str(args.load_env))
+        text = text.replace(GIVEN_NODE, node_info)
+        text = text.replace(COMMAND_SUFFIX, "")
```
[the remark this answers is not in the document]
Comment (pengzhenghao) [reply doc/source/cluster/slurm.rst]: Removed.

**Comment by richardliaw** [comment]
Hey @pengzhenghao just an update, I'm currently working on a draft of edits locally and will push something by tomorrow to this branch!

**Comment by richardliaw** [comment]
I've pushed updates to this branch!

@sumanthratna @pengzhenghao please take a look? I've basically expanded on your example and added a walkthrough guide. Let me know if there's anything I can improve!

**Comment by richardliaw** [comment]
Please take a look by tomorrow and let me know if you have any comments/concerns :) Otherwise I will merge it in, and we can open followup PRs to address comments.

File: doc/source/cluster/examples/slurm-basic.sh [hunk 4]
```diff
@@ -0,0 +1,65 @@
+#!/bin/bash
```
Comment (sumanthratna) [inline doc/source/cluster/examples/slurm-basic.sh:1]: I think this needs to be regenerated? not sure; seems to be a little out of sync w/ slurm-template.sh

Comment (richardliaw) [reply doc/source/cluster/examples/slurm-basic.sh:1]: purposely, it's a separate script (not autogenerated)

File: doc/source/cluster/examples/slurm-template.sh [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+if [[ $ip == *" "* ]]; then
+  IFS=' ' read -ra ADDR <<<"$ip"
+  if [[ ${#ADDR[0]} -gt 16 ]]; then
+    ip=${ADDR[1]}
+  else
+    ip=${ADDR[0]}
+  fi
+  echo "We detect space in ip! You are using IPV6 address. We split the IPV4 address as $ip"
+fi
```
Comment (sumanthratna) [inline doc/source/cluster/examples/slurm-template.sh:41]:
```suggestion
# if we detect a space character in the head node IP, we'll
# convert it to an ipv4 address. This step is optional.
if [[ "$ip" == *" "* ]]; then
  IFS=' ' read -ra ADDR <<< "$ip"
  if [[ ${#ADDR[0]} -gt 16 ]]; then
    ip=${ADDR[1]}
  else
    ip=${ADDR[0]}
  fi
  echo "IPV6 address detected. We split the IPV4 address as $head_node_ip"
fi
```
(just copying over comments from other file)

**Review by sumanthratna — APPROVED** [review]
I'm pretty busy right so I can't take a detailed look, but this looks awesome!

File: doc/source/cluster/slurm.rst [hunk 6]
```diff
@@ -3,7 +3,188 @@
...
+
+.. code-block:: bash
+
+    python slurm-launch.py --exp-name test --command "python your_file.py" --num-nodes 3
+
+If you want to specify the computing nodes, just use the same node name in ``sinfo`` command:
+
+.. code-block:: bash
+
+    python slurm-launch.py --exp-name test --command "python your_file.py" --num-nodes 3 --node chpc-cn[003-005]
```
Comment (pengzhenghao) [inline doc/source/cluster/slurm.rst:164]: Maybe change the name `chic-cn[003-005]` to a placeholder like `NODE_NAMES` or so would be better.

File: doc/source/cluster/slurm.rst [hunk 7]
```diff
@@ -3,7 +3,188 @@
...
+Usage example
+~~~~~~~~~~~~~
+
+If you want to utilize a multi-node cluster in slurm:
+
+.. code-block:: bash
+
+    python slurm-launch.py --exp-name test --command "python your_file.py" --num-nodes 3
+
+If you want to specify the computing nodes, just use the same node name in ``sinfo`` command:
```
Comment (pengzhenghao) [inline doc/source/cluster/slurm.rst:160]:
```suggestion
If you want to specify the computing node(s), just use the same node name(s) in the same format of the output of ``sinfo`` command:
```

**Comment by richardliaw** [comment]
@pengzhenghao thanks a bunch for getting this started!

**Comment by cupdike** [comment]
Where can I find simple-traininer.py?  I can't find it at the location mentioned (or anywhere)...

```
# ray/doc/source/cluster/examples/simple-trainer.py
python -u simple-trainer.py "$SLURM_CPUS_PER_TASK"
```

**Comment by pengzhenghao** [comment]
> @pengzhenghao Where can I find simple-traininer.py? I can't find it at the location mentioned (or anywhere)...
> 
> ```
> # ray/doc/source/cluster/examples/simple-trainer.py
> python -u simple-trainer.py "$SLURM_CPUS_PER_TASK"
> ```

Hi @cupdike , I think this file is what you looking for:

https://github.com/pengzhenghao/ray/blob/c41c2a299f5ec1d53ff5a763b6932f85299ae134/doc/source/cluster/examples/simple-trainer.py

**Pull request merged**